frame/offences/benchmarking/src/lib.rs

```diff
@@ -151,25 +178,162 @@ benchmarks! {
 			validator_set_count: keys.len() as u32,
 			offenders,
 		};
-
+		assert_eq!(System::<T>::event_count(), 0);
 	}: {
 		let _ = <T as ImOnlineTrait>::ReportUnresponsiveness::report_offence(reporters, offence);
 	}
+	verify {
+		// make sure the report was not deferred
+		assert!(Offences::<T>::deferred_offences().is_empty());
+		// make sure that all slashes have been applied
+		assert_eq!(
+			System::<T>::event_count(), 0
+			+ 1 // offence
+			+ 2 * r // reporter (reward + endowment)
+			+ o // offenders slashed
+			+ o * n // nominators slashed
+		);
```

Comment:
This looks a bit hard and flaky. If you know the exact type of the event, how about asserting on `.filter().count()` (i.e. ensure x events of type y have been submitted).

Reply:
It definitely is, but asserting this allowed me to figure out quite many broken assumptions in the original code (i.e. no slashing was taking place at all because of multitude of reasons).
I feel like asserting concrete types will break exactly under the same circumstances as just the number, but adds an extra complexity here.
Currently the `frame_system::Trait::Event` type is set to `()` so we actually don't store any concrete event details at all.

Reply:
Fair enough. As long as you don't see a way for the count to stay the same whilst events being different, then it should be good.

Reply:
This test is indeed flaky... or at least the setup is.

I tried to change the raw_amount to some generic value and the number of events started to change.

We cannot use the value 1_000_000 in the setup here, because that may not even be the existential deposit of a chain.

How did you debug the number of events here?

It seems the initial case is working, but I get this:

```
---- tests::test_benchmarks stdout ----
thread 'tests::test_benchmarks' panicked at 'assertion failed: `(left == right)`
  left: `1701`,
 right: `1901`', frame/offences/benchmarking/src/lib.rs:188:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Reply:
We should instead use `T::Currency::minimum_balance()`

Reply:
Also, in general, this is a lot of events to trigger for one function. This cant be good if we allow something in the runtime to explode so much.

Reply:
Added strict matching on events for `im-online` case (since currently this is the only one that we trigger anyway). It introduces some extra complexity, but might be worth it cause it's much easier to debug now.